### DISCUSSION TITLE
Wild * file name on command line

### DISCUSSION THREAD

**Discussion by jgsuk** [sumatrapdfreader/sumatrapdf#3008] [Q&A]
From my application, I 'shell' to Sumatra to print drawing files (PDFs).
Is it possible to use a 'wild card' for the file names? e.g.

SumatraPDF-3.4.6-32.exe -print-to "myprinter" "myfolder\*.pdf"

Thanks,
Geoff

**Comment by GitHubRulesOK** [comment]
hmm not a good idea as the use of such constructs is acceptable in NiX but not WinDOwS where `forfiles` was/is the norm

windows traditional line limit is 256 chars and expanding * fails after a few names especially using console
The handling of PDF file generally needs to be very much one at a time but SummatraPDF can stack the sequential requests quickly
The limiting factor in printing is the copious amount of resources needed for image processing so chunking as spooled outputs is preferable.

With all that in mind it depends a lot on printer driver settings, the lower the resolution the faster/leaner the print demands.

Anyway I digressed for plotting `dwg1.pdf dwg2.pdf dwg3.pdf ...` a single command line would suffice however for a folder of long names it will usually fail, so simpler script the drag and drop, could be one line but  test with a bat to see the feed back /progress

PdfFolder.bat
```
@echo off& title Run my command in (sub)folder(s)
REM ensure we are running at top level dir
 cd /d "%~1"
 for /R %%f in (*.pdf) do @sumatrapdf [options] "%%f" 
 echo/
pause for effect
```
`sumatrapdf ` should really be the fully qualified location `"c:\program files\sumatrapdf\sumatrapdf.exe"`

for one line you could use forfiles or mod above to `cd /d "foldername" & for /R %%f in (*.pdf) do @sumatrapdf [options] "%%f"`

**Reply by jgsuk** [reply]
Thanks, I thought that might be the case.

The problem that I have is that I am printing a series of drawings, and the order is important in terms of constructing a product.
What appears to be happening (I am shelling to sumatra for each PDF) is that prints sometimes get out of order.  Could that happen if files are of different sizes, with a large file taking a while to process, and a small file 'jumping the queue'.

Regards,
Geoff

**Reply by GitHubRulesOK** [reply]
likely as its down to system driver priority settings but out of order is a common statement if using * since often the sequence is determined by dates or human illogical lexical namesort 1, 10, 11, 2, 20, 3,

**Reply by jgsuk** [reply]
I think that I'll construct a batch file:

Sumatra File1
Sumatra File2
etc.

and then execute that.
The second file print won't then start until File1 has been spooled.
I would think!

Geoff

**Reply by GitHubRulesOK** [reply]
SumatraPDF simply hands data over to spooler ASAP so its spooler sets priorities

if time to run in background is fairly constant then you can add a timeout such that there is only one/two queuing at a time
 so if takes 60 seconds each use a 59/58 second delay 
that forces one in progress and only one other in the wings

**Reply by jgsuk** [reply]
My other idea was to merge the PDF files, and then just do one print.  Do you have any suggestions re merging software? - accessible from within my application, not an interactive, or web based.

Thanks,
Heoff

**Reply by GitHubRulesOK** [reply]
CLI merge is variable for say bookmarks but you just want sequential page rendering, 

Single exe prefered, so most will do, technically single line tools include MuPDF Mutool merge, (compatible with SumatraPDF) no print only file output or its sister Ghostscript but thats a bigger bundle capable of better direct printing support, 

cpdf.exe good for alterations such as n-up, has a community version but cant print
qpdf.exe needs a few dlls (Open Source  allows for commercial use) again good for merge but no print function.

For merging the rendered pages as images you can of course convert to PNG by say xpdf.exe/poppler (FOSS version needs few dlls) then zip/tar as "cbZ" then use SumatraPDF to print the zip.cbz more steps but in some cases may be preferred

OR you could replace SumatraPDF by a better print orientated tool Like Acrobat Reader Print or Tracker eXchange Viewer (modified silent view and print) or really a more dedicated pdf2prn solution like pdf2print(er)

**Comment by kjk** [comment]
Sumatra doesn't recognize wildcard file names.
